### ISSUE TITLE
Add support for Pull Request Reviews and Review Requests

### ISSUE THREAD

**Issue by m-deck** [slapperwan/gh4a#391]
https://developer.github.com/v3/pulls/reviews/

**Comment by sbrl** [comment]
This would be really useful as a maintainer of a project.

**Comment by Tunous** [comment]
Out of preview https://developer.github.com/changes/2017-05-09-end-black-cat-preview/

**Comment by Tunous** [comment]
![spectacle x29041](https://user-images.githubusercontent.com/5156340/26896947-f896d98c-4bc6-11e7-9d73-93b5f6b95dcf.png)

Some in progress work. What you can see here is review groups with matching diffs displayed similar to how GitHub does it. (These diffs would display few lines of code) It's really messy right now but what do you think about this?

In finished version, I think it would be nice to group review blocks into cards to make it easy to see where they start and end. Similar how it works on GitHub mobile web version. And perhaps could add some way to collapse/expand them.

**Comment by sbrl** [comment]
@Tunous That looks great! And displaying them as cards is a fantastic idea. It doesn't have to display things _identically_ to the web version, because sometimes other display methods make more sense.

**Comment by maniac103** [comment]
@Tunous I think I'd prefer displaying a short summary ('Tunous reviewed on xxx\nCode comments in file yyy, zzz\nShow details') and show the details in a separate event list fragment in a new activity. Otherwise I think the list will become either confusing/hard to parse (your prototype) and/or visually jarring (nested list).

**Comment by Tunous** [comment]
![peek 2017-06-10 13-22](https://user-images.githubusercontent.com/5156340/27002344-e37d2234-4ddf-11e7-9c04-59b889017711.gif)

Something like this?

**Comment by Tunous** [comment]
![spectacle m11874](https://user-images.githubusercontent.com/5156340/27002364-75c1fbf6-4de0-11e7-880d-f453bad5b156.png)

With reply buttons of course. Commenting here is less linear so it should work differently:
Bottom sheet would be disabled by default here and only activate once you press one of the reply. And the reply button would be highlighted so you see to which group you are replying.

**Comment by Tunous** [comment]
![peek 2017-06-10 13-32](https://user-images.githubusercontent.com/5156340/27002395-62e1c4de-4de1-11e7-8980-e754b071ed76.gif)

Example with real conversations from [this pr](https://github.com/Kotlin/KEEP/pull/71). You can test it yourself in reviews branch: https://github.com/Tunous/gh4a/tree/reviews. (Code is really messy right now, I'll clean up once everything works)

**Comment by maniac103** [comment]
Looks good. I assume the big green 'diff' block is a placeholder?

**Comment by Tunous** [comment]
Yes, I plan on displaying it like GitHub does (line numbers, text, red/green bgs)

**Comment by Tunous** [comment]
![spectacle m15193](https://user-images.githubusercontent.com/5156340/27002539-10233c9c-4de5-11e7-971f-ea783193e2fc.png)

Like this but I need to figure out how to style it with text spans.

**Comment by maniac103** [comment]
#630 is merged now -> closing

**Comment by Tunous** [comment]
No, not yet. There is also review creation part missing that I'm still working on. Right now we only have a way to view reviews which isn't what a full support should be.

**Comment by maniac103** [comment]
OK, nevermind then :-)

**Comment by smichel17** [comment]
Would it make sense to pin the reply bar to the bottom in the review view, instead of an action at the bottom?

**Comment by Tunous** [comment]
No, there can be multiple comment threads in a single review and each of them has its own reply button.

**Comment by Tunous** [comment]
Actually it would be possible if we did it like in my other idea. (I don't remember where I wrote it)

Reply button would work as a focus point. If you click it, it gets highlighted to tell you that you are replying to its comments thread. At any time you could click on other reply buttons to change the focus. (accidently clicked wrong one?)

That way bottom sheet would be always visible and you would have ability to scroll through all the comments, quote them, copy text, etc. 

I started work on this approach before but I don't remember why I changed my mind.... Probably because comment bottom sheet wasn't merged yet and I didn't want to wait for it.

**Comment by smichel17** [comment]
The way I was thinking about it when I made the suggestion was that:

- On the PR conversation tab, show "Q reviewed this".
- Tapping the review *expands* it to shows the comment threads (like the github web interface), rather than opening a new screen.
- Each comment thread is a card that shows the file (title), a very abbreviated diff, the top comment, and a little bit of additional info ("X more comments").
- Tapping the card opens up a new view that shows more of the diff, all of the comments, and has the reply box pinned to the bottom

**Comment by Tunous** [comment]
I've added checklist to the issue body to list exactly what is done and what else needs to be done.
